### PR TITLE
POL-1405 Replace deprecated Kubecost endpoints for Kubecost Cluster Policy

### PR DESCRIPTION
**Pull request by midir99**
### Description

The endpoint we are currently using has been deprecated “/model/savings/clusterSizing"
It still works but only returns one cluster rather than all clusters.

The new endpoint is returns a different response “/model/savings/clusterSizingETL" and simply adding ETL onto the end of our current end point in the policy returns null values in ds_cluster_sizing

This endpoint returns all cluster recommendations in one response.

The `ds_clusters` function has been created  to include two new parameters in the Kubecost API call:

* param_allow_shared_core: This boolean parameter indicates whether shared cores should be allowed in the cluster sizing calculation. By default, Kubecost may not consider the shared cores between pods as part of the cost savings calculations. With this new parameter, the explicit inclusion of shared cores in the calculations can be enabled. This is useful for environments where resources are shared between multiple pods, and a more accurate estimate of potential savings is desired.

#### Changes Made:

* Updated the function responsible for the API call to /model/savings/clusterSizing 
* Added the param_allow_shared_core to the `ds_clusters` function.
* Updated the HTTP call to send these parameters in the Request.
* Set the default values of these parameters to False, meaning shared cores will not be allowed.

> How the response for the new endpoint differs from what we initially wanted: we modified it and added new functions to handle the response and convert it to the expected struct

### Issues Resolved

- Deprecated endpoint from kubecost was replaced.
- https://flexera.atlassian.net/browse/POL-1405

### Link to Example Applied Policy

https://app.flexeratest.com/orgs/1105/automation/applied-policies/projects/60073?policyId=673520d66168ce351f89f548

### Contribution Check List

- [ ] New functionality includes testing.
- [ ] New functionality has been documented in the README if applicable
- [X] New functionality has been documented in CHANGELOG.MD

### CODE REVIEW

File: cost/kubecost/cluster/kubecost_cluster_rightsizing_recommendations.pt [hunk 1]
```diff
@@ -86,16 +86,6 @@ end
 # Datasources & Scripts
 ###############################################################################
 
-# Get applied policy metadata for use later
-datasource "ds_applied_policy" do
```
[the remark this answers is not in the document]
Comment (joseluiszuflores) [reply cost/kubecost/cluster/kubecost_cluster_rightsizing_recommendations.pt:90]: Hi @XOmniverse  yes you are right looks like we already returned this datasource

File: cost/kubecost/cluster/CHANGELOG.md [hunk 2]
```diff
@@ -1,5 +1,9 @@
 # Changelog
 
+## v0.3.2
+
+- Replace usage of Kubecost deprecated endpoints for newest versions
```
Comment (XOmniverse) [inline cost/kubecost/cluster/CHANGELOG.md:5]: This is awkwardly worded. Recommended change:

- Updated API endpoints for Kubecost. Functionality unchanged.

File: cost/kubecost/cluster/kubecost_cluster_rightsizing_recommendations.pt [hunk 3]
```diff
@@ -162,79 +162,114 @@ script "js_target_util", type:"javascript" do
...
+datasource "ds_get_cluster_list" do
+  run_script $js_get_cluster_list, $ds_clusters
+end
+
+script "js_get_cluster_list", type: "javascript" do
+  parameters "ds_clusters"
+  result "result"
+  code <<-EOS
+  result = []
+_.each(ds_clusters.accountID, function(value, key){
```
Comment (XOmniverse) [inline cost/kubecost/cluster/kubecost_cluster_rightsizing_recommendations.pt:190]: Code should be tabbed out similar to what is seen on line 189

File: cost/kubecost/cluster/kubecost_cluster_rightsizing_recommendations.pt [hunk 4]
```diff
@@ -291,12 +329,7 @@ end
 
 policy "pol_kubecost_recommendations" do
   validate_each $ds_recommendations do
-    summary_template "{{ with index data 0 }}{{ .policy_name }}{{ end }}"
-    detail_template <<-'EOS'
```
Comment (XOmniverse) [inline cost/kubecost/cluster/kubecost_cluster_rightsizing_recommendations.pt:295]: Why are we removing this? What does this have to do with updating API endpoints?

File: cost/kubecost/cluster/kubecost_cluster_rightsizing_recommendations.pt [hunk 5]
```diff
@@ -248,40 +283,43 @@ script "js_recommendations", type: "javascript" do
...
-    policy_name: ds_applied_policy['name'],
     min_nodes: param_min_nodes,
     lookback: param_lookback,
     target_util: param_target_util,
     message: message,
+    policy_name: ds_applied_policy['name'],
     message_savings: [
       ds_currency['symbol'], ' ',
-      formatNumber(Math.round(recommendation['monthlySavings'] * -1000) / 1000, ds_currency['separator'])
+      formatNumber(Math.round(recommendation['monthlySavings']).toFixed(2), ds_currency['separator'])
```
Comment (XOmniverse) [inline cost/kubecost/cluster/kubecost_cluster_rightsizing_recommendations.pt:318]: Why is this change being made? Removing a multiplication against a negative number is going to change the result. Is this intended?

**Review by XOmniverse — CHANGES_REQUESTED** [review]
I requested some specific changes, but more broadly, there appear to be a lot more changes here than just updating deprecated API endpoints.

Please fully document all changes in the changelog.

File: cost/kubecost/cluster/kubecost_cluster_rightsizing_recommendations.pt [hunk 6]
```diff
@@ -162,79 +162,114 @@ script "js_target_util", type:"javascript" do
...
-    field "totalNodeCount", jmes_path(response, "data.currentClusterInfo.totalCounts.totalNodeCount")
-    field "totalRAMGB", jmes_path(response, "data.currentClusterInfo.totalCounts.totalRAMGB")
-    field "totalVCPUs", jmes_path(response, "data.currentClusterInfo.totalCounts.totalVCPUs")
-    field "monthlyRate", jmes_path(response, "data.currentClusterInfo.monthlyRate")
-    field "recommendations", jmes_path(response, "data.recommendations")
+    field "accountID", jmes_path(response, "data")
   end
 end
 
+datasource "ds_get_cluster_list" do
```
Comment (nia-vf1) [inline cost/kubecost/cluster/kubecost_cluster_rightsizing_recommendations.pt:181]: Where is this datasource used? If not used, then it should be removed from the template along with the `js_get_cluster_list` script

Comment (joseluiszuflores) [reply cost/kubecost/cluster/kubecost_cluster_rightsizing_recommendations.pt]: done

File: cost/kubecost/cluster/kubecost_cluster_rightsizing_recommendations.pt (hunk 5 above)
Comment (joseluiszuflores) [reply cost/kubecost/cluster/kubecost_cluster_rightsizing_recommendations.pt]: this was test code. I reverted this one  now the behaviur should be same

File: cost/kubecost/cluster/kubecost_cluster_rightsizing_recommendations.pt (hunk 4 above)
Comment (joseluiszuflores) [reply cost/kubecost/cluster/kubecost_cluster_rightsizing_recommendations.pt]: I reverted this one also

File: cost/kubecost/cluster/kubecost_cluster_rightsizing_recommendations.pt (hunk 3 above)
Comment (joseluiszuflores) [reply cost/kubecost/cluster/kubecost_cluster_rightsizing_recommendations.pt]: Done I have followed the indentation

File: cost/kubecost/cluster/CHANGELOG.md [hunk 7]
```diff
@@ -1,5 +1,11 @@
 # Changelog
 
+## v0.3.3
+
+- Updated API endpoints for Kubecost.
+- Added two new params param_allow_shared_core,param_include_overhead.
```
Comment (Rajasundar) [inline cost/kubecost/cluster/CHANGELOG.md:6]: we can use backticks or quotes here around the parameter names to indicate they are code or field names.

File: cost/kubecost/cluster/kubecost_cluster_rightsizing_recommendations.pt [hunk 8]
```diff
@@ -163,35 +180,71 @@ script "js_target_util", type:"javascript" do
...
+end
+
 datasource "ds_recommendations" do
-  run_script $js_recommendations, $ds_cluster_sizing, $ds_currency, $ds_applied_policy, $param_strategy, $param_min_nodes, $param_lookback, $param_target_util
+  run_script $js_recommendations, $ds_cluster_sizing, $ds_currency, $ds_applied_policy,$param_strategy, $param_min_nodes, $param_lookback, $param_target_util
 end
 
 script "js_recommendations", type: "javascript" do
-  parameters "ds_cluster_sizing", "ds_currency", "ds_applied_policy", "param_strategy", "param_min_nodes", "param_lookback", "param_target_util"
+  parameters "ds_cluster_sizing", "ds_currency","ds_applied_policy", "param_strategy", "param_min_nodes", "param_lookback", "param_target_util"
```
Comment (Rajasundar) [inline cost/kubecost/cluster/kubecost_cluster_rightsizing_recommendations.pt:244]: Inconsistency in spacing. 
```suggestion
  parameters "ds_cluster_sizing", "ds_currency", "ds_applied_policy", "param_strategy", "param_min_nodes", "param_lookback", "param_target_util"
```

File: cost/kubecost/cluster/kubecost_cluster_rightsizing_recommendations.pt [hunk 9]
```diff
@@ -163,35 +180,71 @@ script "js_target_util", type:"javascript" do
...
+      monthlyRate: accountData.currentClusterInfo.monthlyRate,
+      recommendations: accountData.recommendations
+    });
+  }
+  EOS
+end
+
 datasource "ds_recommendations" do
-  run_script $js_recommendations, $ds_cluster_sizing, $ds_currency, $ds_applied_policy, $param_strategy, $param_min_nodes, $param_lookback, $param_target_util
+  run_script $js_recommendations, $ds_cluster_sizing, $ds_currency, $ds_applied_policy,$param_strategy, $param_min_nodes, $param_lookback, $param_target_util
```
Comment (Rajasundar) [inline cost/kubecost/cluster/kubecost_cluster_rightsizing_recommendations.pt:240]: Inconsistency in spacing.

```suggestion
  run_script $js_recommendations, $ds_cluster_sizing, $ds_currency, $ds_applied_policy, $param_strategy, $param_min_nodes, $param_lookback, $param_target_util
```

**Review by Rajasundar — CHANGES_REQUESTED** [review]
@joseluiszuflores, I've added few comments with questions for clarification.

File: cost/kubecost/cluster/kubecost_cluster_rightsizing_recommendations.pt [hunk 10]
```diff
@@ -36,6 +36,23 @@ parameter "param_kubecost_host" do
...
+
+parameter "param_include_overhead" do
+  type "string"
+  category "Include overhead"
+  label "Include overhead"
+  allowed_values "true", "false"
+  description "By enabling this option, Kubecost takes into account not only the resources used by applications and containers, but also those needed to operate and maintain the Kubernetes environment"
+end
+
+
```
Comment (nia-vf1) [inline cost/kubecost/cluster/kubecost_cluster_rightsizing_recommendations.pt:55]:
```suggestion
```

**Review by Rajasundar — DISMISSED** [review]
LGTM

File: cost/kubecost/cluster/CHANGELOG.md [hunk 11]
```diff
@@ -1,5 +1,11 @@
 # Changelog
 
+## v0.3.3
+
+- Updated API endpoints for Kubecost.
+- Added two new params `param_allow_shared_core`, `param_include_overhead`.
```
Comment (TDubovchenko) [inline cost/kubecost/cluster/CHANGELOG.md:6]: @joseluiszuflores this line does not help users to understand the change

File: cost/kubecost/cluster/kubecost_cluster_rightsizing_recommendations.pt [hunk 12]
```diff
@@ -36,6 +36,22 @@ parameter "param_kubecost_host" do
   # No default value, user input required
 end
 
+parameter "param_allow_shared_core" do
+  type "string"
+  category "Allow Shared Core"
+  label "Shared core"
+  allowed_values "true", "false"
+  description "Enable or disable the use of shared cores in calculating costs and resource savings in Kubecost"
```
Comment (TDubovchenko) [inline cost/kubecost/cluster/kubecost_cluster_rightsizing_recommendations.pt:44]: @joseluiszuflores could we re-phrase the description to something like: "Whether to include shared core node types into recommendation."

File: cost/kubecost/cluster/kubecost_cluster_rightsizing_recommendations.pt [hunk 13]
```diff
@@ -36,6 +36,22 @@ parameter "param_kubecost_host" do
...
+end
+
+parameter "param_include_overhead" do
+  type "string"
+  category "Include overhead"
+  label "Include overhead"
+  allowed_values "true", "false"
+  description "By enabling this option, Kubecost takes into account not only the resources used by applications and containers, but also those needed to operate and maintain the Kubernetes environment"
```
Comment (TDubovchenko) [inline cost/kubecost/cluster/kubecost_cluster_rightsizing_recommendations.pt:52]: We might want to improve the description here.

File: cost/kubecost/cluster/CHANGELOG.md (hunk 11 above)
Comment (joseluiszuflores) [reply cost/kubecost/cluster/CHANGELOG.md]: we have changed log file with the next :

## v0.3.3

- Updated from /model/savings/clusterSizing to /model/savings/clusterSizingETL.
- Created new functions to handle the modified response format from the updated endpoint.
- Modified functionality to generate incident correctly based on the new response structure.
- New Parameters:
  - param_allow_shared_core: Specifies whether shared cores should be considered in the cluster sizing calculation.
  - param_include_overhead: Ensures overhead costs are correctly included in the cost calculations.

File: cost/kubecost/cluster/kubecost_cluster_rightsizing_recommendations.pt (hunk 12 above)
Comment (joseluiszuflores) [reply cost/kubecost/cluster/kubecost_cluster_rightsizing_recommendations.pt]: yeah we have modified  this one by  ***Whether to include shared core node types into recommendation.***

File: cost/kubecost/cluster/kubecost_cluster_rightsizing_recommendations.pt (hunk 13 above)
Comment (joseluiszuflores) [reply cost/kubecost/cluster/kubecost_cluster_rightsizing_recommendations.pt]: We have changed by one *** Whether to include resources for operating and maintaining the Kubernetes environment, in addition to those used by applications and containers.***

We don't have much information about overhead but we saw before on Kubecost UI send this flag as false in order to have coherence with UI and  our incidents  we decide to add this flag. Kubecost said this about this concept: [link](https://docs.kubecost.com/using-kubecost/navigating-the-kubecost-ui/assets)

> Overhead costs refer to the capacity costs minus allocatable costs. This includes Kubelet, container runtime, and node operating systems and software. It does not include Prometheus, DNS, cert-manager, kube-system, or pods running with Kubernetes. To learn more about overhead costs, watch [Node Size Matters](https://www.youtube.com/watch?v=6vNI_O6sdvY), a presentation by Kubecost engineers from KubeCon 2023.

And on the video mentioned, talks about Overhead as: 

....  overhead in the Kubernetes environment often refers to the resources consumed beyond what is directly allocated to user applications, such as maintaining the cluster and running system components.

Comment (joseluiszuflores) [reply cost/kubecost/cluster/kubecost_cluster_rightsizing_recommendations.pt]: done we modified  the description

File: cost/kubecost/cluster/kubecost_cluster_rightsizing_recommendations.pt [hunk 14]
```diff
@@ -36,6 +36,24 @@ parameter "param_kubecost_host" do
...
+  category "Allow Shared Core"
+  label "Shared core"
+  description "Whether to include shared core node types into recommendation."
+  allowed_values "true", "false"
+  default "false"
+end
+
+parameter "param_include_overhead" do
+  type "string"
+  category "Include overhead"
```
[the remark this answers is not in the document]
Comment (joseluiszuflores) [reply cost/kubecost/cluster/kubecost_cluster_rightsizing_recommendations.pt:50]: done!

File: cost/kubecost/cluster/kubecost_cluster_rightsizing_recommendations.pt [hunk 15]
```diff
@@ -36,6 +36,24 @@ parameter "param_kubecost_host" do
...
+  label "Include overhead"
```
[the remark this answers is not in the document]
Comment (joseluiszuflores) [reply cost/kubecost/cluster/kubecost_cluster_rightsizing_recommendations.pt:51]: Done! Not neccesary because include "overhead" was removed

File: cost/kubecost/cluster/kubecost_cluster_rightsizing_recommendations.pt [hunk 16]
```diff
@@ -36,6 +36,24 @@ parameter "param_kubecost_host" do
...
+  description "Enabling this option lets Kubecost track the resources used by both your applications and the Kubernetes infrastructure."
```
[the remark this answers is not in the document]
Comment (joseluiszuflores) [reply cost/kubecost/cluster/kubecost_cluster_rightsizing_recommendations.pt:52]: Done! Not necessary as include overhead has been removed.

File: cost/kubecost/cluster/README.md [hunk 17]
```diff
@@ -12,6 +12,8 @@ This policy template reports a Kubecost cluster rightsizing recommendation gener
 - *Lookback Period (Days)* - Number of historical days of usage to analyze when generating recommendations.
 - *Target Utilization (%)* - Utilization target to use when generating recommendations.
 - *Recommendation Strategy* - Recommendation strategy to use. 'Optimal' will automatically select whichever strategy has the highest potential savings.
+- *Allow Shared Core* Specifies whether shared cores should be considered in the cluster sizing calculation.
```
[the remark this answers is not in the document]
Comment (joseluiszuflores) [reply cost/kubecost/cluster/README.md:15]: done!

File: cost/kubecost/cluster/kubecost_cluster_rightsizing_recommendations.pt [hunk 18]
```diff
@@ -36,6 +36,24 @@ parameter "param_kubecost_host" do
   # No default value, user input required
 end
 
+parameter "param_allow_shared_core" do
+  type "string"
+  category "Allow Shared Core"
+  label "Shared core"
+  description "Whether to include shared core node types into recommendation."
```
[the remark this answers is not in the document]
Comment (joseluiszuflores) [reply cost/kubecost/cluster/kubecost_cluster_rightsizing_recommendations.pt]: done!

File: cost/kubecost/cluster/CHANGELOG.md [hunk 19]
```diff
@@ -1,5 +1,12 @@
 # Changelog
 
+## v0.3.3
+
+- Updated API endpoints for cluster Sizing.
+- Added new parameters `param_allow_shared_core`, `param_include_overhead`:
```
[the remark this answers is not in the document]
Comment (joseluiszuflores) [reply cost/kubecost/cluster/CHANGELOG.md]: done!

File: cost/kubecost/cluster/README.md [hunk 20]
```diff
@@ -12,6 +12,8 @@ This policy template reports a Kubecost cluster rightsizing recommendation gener
...
+- *Include Overhead* Ensures that overhead costs, such as control plane and node management resources, are included in the cost calculations.
```
[the remark this answers is not in the document]
Comment (joseluiszuflores) [reply cost/kubecost/cluster/README.md]: done !

File: cost/kubecost/cluster/kubecost_cluster_rightsizing_recommendations.pt [hunk 21]
```diff
@@ -163,27 +180,48 @@ script "js_target_util", type:"javascript" do
...
   request do
     host $param_kubecost_host
-    path "/model/savings/clusterSizing"
+    path "/model/savings/clusterSizingETL"
     query "minNodeCount", $param_min_nodes
     query "window", join([$param_lookback, "d"])
     query "targetUtilization", val($ds_target_util, "value")
+    query "allowSharedCore", $param_allow_shared_core
+    query "architecture", $param_architecture
+    query "includeOverhead", "false"
```
Comment (Rajasundar) [inline cost/kubecost/cluster/kubecost_cluster_rightsizing_recommendations.pt:192]: let's keep a note here adding why the `includeOverhead` field is included and set to `false`

Comment (joseluiszuflores) [reply cost/kubecost/cluster/kubecost_cluster_rightsizing_recommendations.pt:192]: done!

File: cost/kubecost/cluster/kubecost_cluster_rightsizing_recommendations.pt [hunk 22]
```diff
@@ -216,74 +254,79 @@ script "js_recommendations", type: "javascript" do
...
+      "The above settings can be modified by editing the applied policy and changing the appropriate parameters."
     ].join('')
-  }]
-EOS
+    savings = (Math.round(recommendation['monthlySavings'] * 100) / 100 ).toFixed(2)
+
+    recommendationDetails = [
+      "Modify settings for cluster ", accountID,
+      " so that node count is set to ", recommendation['nodeCount'],
+      " and pools are configured to match the Recommended Pools field."
```
[the remark this answers is not in the document]
Comment (joseluiszuflores) [reply cost/kubecost/cluster/kubecost_cluster_rightsizing_recommendations.pt]: We have changed recommendationDetails

File: cost/kubecost/cluster/kubecost_cluster_rightsizing_recommendations.pt [hunk 23]
```diff
@@ -163,27 +180,49 @@ script "js_target_util", type:"javascript" do
...
   request do
     host $param_kubecost_host
-    path "/model/savings/clusterSizing"
+    path "/model/savings/clusterSizingETL"
     query "minNodeCount", $param_min_nodes
     query "window", join([$param_lookback, "d"])
     query "targetUtilization", val($ds_target_util, "value")
+    query "allowSharedCore", $param_allow_shared_core
+    query "architecture", $param_architecture
+    # We set "includeOverhead" to "false" because omitting it causes mismatched results on recommendations
```
Comment (Rajasundar) [inline cost/kubecost/cluster/kubecost_cluster_rightsizing_recommendations.pt:192]:
```suggestion
    # Note: `includeOverhead` is explicitly set to `false` to ensure consistency between the recommendations and the Kube-cost UI. 
    # Also, the field  `includeOverhead`  is not exposed as a configurable parameter because it is not part of the Cluster Right-Sizing Recommendation API specification.
```

@joseluiszuflores, I'll leave the decision to you.

**Review by Rajasundar — APPROVED** [review]

**Review by XOmniverse — CHANGES_REQUESTED** [review]
Minor changes. Also, merge conflicts need to be resolved.

File: cost/kubecost/cluster/kubecost_cluster_rightsizing_recommendations.pt [hunk 24]
```diff
@@ -36,6 +36,24 @@ parameter "param_kubecost_host" do
...
+  default "false"
+end
+
+parameter "param_architecture" do
+  type "string"
+  category "Policy Settings"
+  label "Architecture"
+  description "Chipset architecture for the recommended nodes. Leave blank to include all architectures. 'ARM' is currently supported only on AWS clusters."
+  allowed_values "", "x86", "ARM"
+  default ""
```
[the remark this answers is not in the document]
Comment (joseluiszuflores) [reply cost/kubecost/cluster/kubecost_cluster_rightsizing_recommendations.pt]: done!

File: cost/kubecost/cluster/kubecost_cluster_rightsizing_recommendations.pt [hunk 25]
```diff
@@ -7,7 +7,7 @@ category "Cost"
 severity "low"
 default_frequency "weekly"
 info(
-  version: "0.3.2",
+  version: "0.3.3",
```
[the remark this answers is not in the document]
Comment (joseluiszuflores) [reply cost/kubecost/cluster/kubecost_cluster_rightsizing_recommendations.pt]: done!

File: cost/kubecost/cluster/CHANGELOG.md [hunk 26]
```diff
@@ -1,5 +1,11 @@
 # Changelog
 
+## v0.4.0
+
+- Updated API endpoint for cluster Sizing.
```
Comment (XOmniverse) [inline cost/kubecost/cluster/CHANGELOG.md:5]: Sizing should not be capitalized.

**Review by XOmniverse — CHANGES_REQUESTED** [review]
One very minor change.

File: cost/kubecost/cluster/CHANGELOG.md (hunk 26 above)
Comment (joseluiszuflores) [reply cost/kubecost/cluster/CHANGELOG.md]: done!

**Review by XOmniverse — APPROVED** [review]
LGTM
